File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-2.java
class Solution {
    public int findMin(int[] nums) {
        // Initialize pointers to the start and end of the array
        int left = 0;
        int right = nums.length - 1;

        // Binary search loop continues while the search space is greater than 1 element
        while (left < right) {
            // Compute the midpoint to avoid integer overflow
            int mid = (right + left) / 2;

            /*
             Example: nums = [4,5,6,7,0,1,2]
             First iteration:
             left = 0, right = 6, mid = 3 → nums[3] = 7, nums[6] = 2

             Since nums[mid] > nums[right], we know the smallest value
             must be to the right of mid, so we move left = mid + 1
            */

            if (nums[mid] < nums[right]) {
                // The min is in the left part (including mid), shrink the right bound
                right = mid;

                /*
                 Example: nums = [3,4,5,1,2]
                 left = 0, right = 4, mid = 2 → nums[2] = 5, nums[4] = 2
                 5 > 2 → move left = mid + 1 → left = 3
                */
            } else {
                // The min is in the right part (excluding mid), shrink the left bound
                left = mid + 1;

                /*
                 Example: nums = [4,5,6,7,0,1,2]
                 mid = 3 → nums[3] = 7, nums[6] = 2 → 7 > 2
                 So we know min is to the right of mid → left = mid + 1 = 4
                */
            }
        }

        // After the loop ends, left == right, pointing at the smallest element
        return nums[left];

        /*
         For example: nums = [4,5,6,7,0,1,2]
         After loop:
         left = right = 4 → nums[4] = 0 → that's the smallest element
        */
    }
}

File: Data Structures & Algorithms/find-in-mountain-array/submission-2.py
class Solution:
    def findInMountainArray(self, target: int, mountainArr: 'MountainArray') -> int:
        l = 1
        r = mountainArr.length() - 2

        while (l <= r):
            m = (l + r) // 2

            left, mid, right = mountainArr.get(m - 1) , mountainArr.get(m), mountainArr.get(m + 1)

            if left < mid < right:
                l = m + 1
            elif left > mid > right:
                r = m - 1
            else:
                peak = m
                break

        l = 0
        r = peak - 1

        while l <= r:
            m = (l + r) // 2

            val = mountainArr.get(m)
            if val < target:
                l = m + 1
            elif val > target:
                r = m - 1
            else:
                return m
        
        l = peak
        r = mountainArr.length() - 1

        while l <= r:
            m = (l + r) // 2
            
            val = mountainArr.get(m)
            if val > target:
                l = m + 1
            elif val < target:
                r = m - 1
            else:
                return m

        return -1